train.py

```diff
@@ -479,521 +480,532 @@ def mlp_policy_fn(module):
+                cpu_state_dict = {}
+                if args["train_type"] in ["bnb_llama_pro", "hqq_llama_pro"]:
+                    trainable_fsdp_modules =[(n,m) for n,m in model.named_modules() if n.endswith(tuple(new_layer_names))]
+                else:
+                    trainable_fsdp_modules = [(n,m) for n,m in model.named_modules() if n.endswith(('lora_AB', 'dora_layer', 'magnitude_layer'))]
+                for prefix, module in trainable_fsdp_modules:
+                    prefix = (prefix.replace("_fsdp_wrapped_module.", "")
+                                    .replace("_checkpoint_wrapped_module.", "")
+                                    .replace("_offload_wrapped_module.", ""))
+                    if args['verbose']: print(f"Saving {prefix}")
+                    with FSDP.state_dict_type(module, StateDictType.FULL_STATE_DICT, save_policy):
+                        cpu_state_dict = {**cpu_state_dict, **{f"{prefix}.{k}":v for k,v in module.state_dict().items()}}
+                    dist.barrier()
+                    torch.cuda.synchronize()
                 if rank==0:
-                    print("Saving full model weights.")
+                    print("Saving trained LoRA weights.")
                     save_file(cpu_state_dict, os.path.join(args["output_dir"], "model_state_dict.safetensors"))
                     print("Done", rank)
+            else:
+                with FSDP.state_dict_type(model, StateDictType.FULL_STATE_DICT, save_policy):
+                    cpu_state_dict = model.state_dict()
+                    if rank==0:
+                        print("Saving full model weights.")
+                        save_file(cpu_state_dict, os.path.join(args["output_dir"], "model_state_dict.safetensors"))
+                        print("Done", rank)
 
-    dist.barrier() # Stop other processes ending while model saving - probably not needed?
+        dist.barrier() # Stop other processes ending while model saving - probably not needed?
 
-    # Clean up
-    dist.destroy_process_group()
+        # Clean up
+        dist.destroy_process_group()
+    if args["profiling_output"]:
+        prof.export_stacks(path = f"{args['profiling_output']}_{local_rank}.txt",
+                           metric = "self_cuda_time_total")
 
 def validate_args(args):
     if args["n_bits"] != 4 and args["train_type"] not in ["hqq_lora", "hqq_dora", "hqq_llama_pro"]:
         raise ValueError(f"train_type={args['train_type']} doesn't support n_bits={args['n_bits']}. Either don't pass n_bits (to use default of 4) or use any of the hqq training types.")
 
-# Entry point, using fastcore's call_parse to parse args from command line and then calling fsdp_main
-@call_parse()
-def main(
+# Main entry point, validate and package arguments, then call fsdp_main. Scripts importing train.py as an import can invoke this function (invoking main() directly leads to call_parse() overwriting arguments).
+def fsdp_qlora(
```

Comment:
Since this no longer uses `call_parse`, could you change the `Param` and `bool_arg` types to normal type hints before merging? And perhaps move the argstrings to a docstring?